### PR TITLE
Replace governance getters with Params and ParamsAt

### PR DESCRIPTION
**Pull request by blukat29**
## Proposed changes

Merging into a feature branch.

1. Update params after inserting/writing each block
- Call `UpdateParams()` in `CreateSnapshot()` to update currentParams (which is returned by `Params()`)
  - CreateSnapshot() is an ideal place to update governance parameters.
    - CreateSnapshot() is called exactly once for each block.
    - A Proposer CN calls CreateSnapshot() in worker.wait(), after writing the mined block
    - Non-proposer CNs, PNs, ENs call CreateSnapshot() in BlockChain.insertChain(), after inserting the new block
- To see how `UpdateParams()` work, see [the code](https://github.com/klaytn/klaytn/blob/feat/contractgov/governance/mixed.go#L66).

2. Replace getters
- Replace 18 governance parameter getters with 2 getters `Params()` and `ParamsAt()`.
- Replaced methods:
  ```go
    GovernanceMode() string
    GoverningNode() common.Address
    UnitPrice() uint64
    CommitteeSize() uint64
    Epoch() uint64
    ProposerPolicy() uint64
    DeferredTxFee() bool
    MinimumStake() string
    MintingAmount() string
    ProposerUpdateInterval() uint64
    Ratio() string
    StakingUpdateInterval() uint64
    UseGiniCoeff() bool
    GetGovernanceValue(key int) interface{}
    GetGovernanceItemAtNumber(num uint64, key string) (interface{}, error)
    GetItemAtNumberByIntKey(num uint64, key int) (interface{}, error)
    GetGoverningInfoAtNumber(num uint64) (bool, common.Address, error)
    GetMinimumStakingAtNumber(num uint64) (uint64, error)
  ```

3. Relocate implementations of `Params()`, `ParamsAt()` and `UpdateParams()` inside default.go.
  - The implementations was originally added in #1380.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

https://github.com/klaytn/klaytn/issues/1375

## Further comments

### CODE REVIEW

File: consensus/istanbul/backend/engine.go [hunk 1]
```diff
@@ -656,8 +651,13 @@ func getPrevHeaderAndUpdateParents(chain consensus.ChainReader, number uint64, h
 
 // CreateSnapshot does not return a snapshot but creates a new snapshot at a given point in time.
 func (sb *backend) CreateSnapshot(chain consensus.ChainReader, number uint64, hash common.Hash, parents []*types.Header) error {
-	_, err := sb.snapshot(chain, number, hash, parents)
-	return err
+	if err := sb.governance.UpdateParams(); err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (blukat29) [reply consensus/istanbul/backend/engine.go:655]: Good point. Fixed.

**Review by ian0371 — APPROVED** [review]

File: governance/default.go [hunk 2]
```diff
@@ -1194,3 +1150,54 @@ func (gov *Governance) IdxCacheFromDb() []uint64 {
...
 }
+
+// Returns the istanbul epoch. This function works even before loading any params
+// from database. We need epoch to load any param from database.
+func (gov *Governance) epochWithFallback() uint64 {
+	// Comes here in most cases
+	if v, ok := gov.Params().Get(params.Epoch); ok {
+		return v.(uint64)
+	}
+	// Useful before first UpdateParam()
```
Comment (yoomee1313) [inline governance/default.go:1161]: Useful..? Does it mean `below if loop is used before first UpdateParam()`? What about changing to more specific sentence?

Comment (blukat29) [reply governance/default.go]: Fixed in https://github.com/klaytn/klaytn/pull/1424/commits/3e859a5b401be6e4d813590a6d7921526a4d59a3.

Comment (yoomee1313) [reply governance/default.go]: Thanks, @blukat29

**Pull request merged**